### ISSUE TITLE
Required CuPy version specification in the stable version looks incomplete

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#4084]
Current minor versions of Chainer requires CuPy (if installed) older than the unseen next minor version. For example, the recent release of Chainer v3.2.0 requires CuPy <2.3. I think it should be fixed so that 1) it does not allow too old CuPy, and 2) it allows any newer version of CuPy at the same major version.

**Comment by kmaehashi** [comment]
I'd like to fix this in next release. For Chainer v3.4.0 release, let's use `2.4<=cupy, cupy<2.5`.

Besides that, we are currently not detecting version mismatch caused after installation (i.e., install cupy, install chainer, then upgrade/downgrade cupy). Runtime detection may also be needed.

**Comment by kmaehashi** [comment]
How about this?

```
For stable branch:
  v4.0.0: `>=4.0.0, <5.0.0`
  v4.1.0: `>=4.1.0, <5.0.0`
  ...
For master branch:
  v5.0.0a1: `==5.0.0a1`
  v5.0.0b1: `==5.0.0b1`
  ...
```

**Comment by kmaehashi** [comment]
PR raised: #4830
(I found the spec has unexpectedly changed in #4640)

**Comment by kmaehashi** [comment]
Fixed via #4830
